learn_evm/arithmetic-checks.md

```diff
@@ -0,0 +1,668 @@
+0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc // ~3
+0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffd // -3 = ~3 + 1
+```
+
+To verify that `-a + a = 0` holds for all integers, we can use the property of two's complement arithmetic that `-a = ~a + 1`.
+By substituting this into the equation, we get `-a + a = (~a + 1) + a = MAX + 1 = 0`, where `MAX` is the maximum integer value.
+
+There is a unique case that needs special attention in two's complement arithmetic.
+The smallest possible integer `int256).min = 0x8000000000000000000000000000000000000000000000000000000000000000 = -57896044618658097711785492504343953926634992332820282019728792003956564819968`
+does not have a positive inverse for addition, making it the only negative number with this property.
+
+Interestingly, if we try to compute `-type(int256).min`, we get the same number, as `-type(int256).min = ~type(int256).min + 1 = type(int256).min`.
+This means there are two fixed points for additive inverses: `-0 = 0` and `-type(int256).min = type(int256).min`.
+It's important to note that Solidity's arithmetic checks will throw an error when evaluating `-type(int256).min` (outside of unchecked blocks).
+
+Finally, looking at the underlying bit (or hex) representation highlights the importance of using the correct operators for signed integers, such as `slt` instead of `lt`, to avoid misinterpreting negative values as large numbers.
+
+```solidity
+  0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff // int256(-1) or type(uint256).max
+< 0x0000000000000000000000000000000000000000000000000000000000000000 // 0
+// When using `slt`, the comparison is interpreted as `-1 < 0 = true`.
+= 0x0000000000000000000000000000000000000000000000000000000000000001
+// When using `lt`, the comparison is interpreted as `type(uint256).max < 0 = false`.
+= 0x0000000000000000000000000000000000000000000000000000000000000000
+```
+
+Newer versions of Solidity prevent integer overflow by using the computed result `c = a + b` to check for overflow/underflow.
+However, unlike unsigned addition, signed addition requires two separate checks instead of one.
+
+```solidity
+/// @notice versions >=0.8.16
+function checkedAddInt2(int256 a, int256 b) public pure returns (int256 c) {
+    unchecked {
+        c = a + b;
+
+        // If `a` is positive, then the sum `c = a + b` can't be less than `b`.
+        if (a > 0 && c < b) arithmeticError();
+        // If `a` is negative, then the sum `c = a + b` can't be greater than `b`.
+        if (a < 0 && c > b) arithmeticError();
+    }
+}
+```
+
+Nevertheless, using the boolean exclusive-or lets us combine these checks into one step.
+Solidity doesn't have a built-in operation for boolean values, but we can still make use of it through inline-assembly. In doing so, we need to take care that both inputs are actually boolean (either 0 or 1), as the xor operation works bitwise and isn't restricted to boolean values.
```

Comment:
I think this is confusing `Solidity doesn't have a built-in operation for boolean values`, as you can use `&&` and `||` on booleans. Maybe solidity doesn't have built-in for boolean xor?

Reply:
Ah, good catch, this was a bit messed up due to my latest push.